### PR TITLE
Include session system

### PR DESCRIPTION
**Pull request by konradhartwig**

### CODE REVIEW

**Comment by lucaschimweg** [comment]
@konradhartwig could you fix codacy review first?

File: src/public/js/app.js [hunk 1]
```diff
@@ -54,8 +80,8 @@ window.addEventListener("load", () => {
         await notificationRoute(el, notificationTemplate, api, showError);
     });
 
-    router.add("/user", async () => {
-        await userRoute(el, logoutTemplate, loginTemplate, api, showError);
+    router.add("/ulogout", async () => {
```
Comment (lucaschimweg) [inline src/public/js/app.js]: would prefer `/logout`. looks way cleaner

File: src/public/index.html [hunk 2]
```diff
@@ -25,9 +25,9 @@
         <a class="item" id="nt" href="/notification">
             Notifications
         </a>
-        <a class="item" href="/user">
-            <i class="user alternate outline icon" href="/user"></i>
-            <p id="user" href="/user"></p>
+        <a class="item" href="/ulogout">
```
Comment (lucaschimweg) [inline src/public/index.html]: prefer `/logout` here as well

File: src/public/js/route/userRoute.js [hunk 3]
```diff
@@ -1,21 +1,26 @@
...
-    //check if logged in or not
-    const user = getUser();
-    let html;
-
-    if(user) {
-        html = logoutTemplate({user: getUser()});
-    } else {
-        html = loginTemplate();
+async function userRoute(el, logoutTemplate, api, showError) {
+    async function logoutFunction() {
```
Comment (lucaschimweg) [inline src/public/js/route/userRoute.js]: please move the inner function out of the outer function. This code looks really confusing.

File: src/server/handler/authMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+export default class AuthMiddleware {
+
+    public static auth(permissionName: string): express.Handler {
+        const namespacedPermission = Config.instance.permissionNamespace + "." + permissionName;
+        return async (req: express.Request, res: express.Response, next: express.NextFunction): Promise<void> => {
+            // get cookie from frontend
+            const sid = req.cookies.HOPPER_SESSION;
+            if (sid == undefined) {
+                console.log("Cookie undefined");
```
Comment (lucaschimweg) [inline src/server/handler/authMiddleware.ts]: do we really want to print this any time someone tries to to open hopper without a session?

File: src/server/handler/authMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                utils.handleError(new Error(Config.instance.authRedirectUrl + "?target=" + Config.instance.baseUrl),
+                    res, 401);
+                return;
+            }
+
+            const session = await Session.decode(sid);
+
+            if (!session){
+                console.log("Cookie not valid");
```
Comment (lucaschimweg) [inline src/server/handler/authMiddleware.ts]: see above

File: src/server/handler/authMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                    res, 401);
+                return;
+            }
+
+            if (!session.user.roles.includes(namespacedPermission)) {
+                utils.handleError(new Error("No permission to use this service"), res, 403);
+                return;
+            }
+
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/authMiddleware.ts]: why is there a ts-ignore here? consider using this: https://github.com/hopperteam/hopper-server/blob/master/src/handler/authMiddleware.ts#L10-L16

File: src/server/handler/notificationHandler.ts [hunk 7]
```diff
@@ -16,16 +16,18 @@ export default class NotificationHandler extends Handler {
 
     private async getAll(req: express.Request, res: express.Response): Promise<void> {
         try {
-            const notifications = await Notification.find({userId: req.query.token}).populate("subscriber");
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/notificationHandler.ts]: why is this here? consider using https://github.com/hopperteam/hopper-server/blob/master/src/handler/authMiddleware.ts#L10-L16

File: src/server/handler/notificationHandler.ts [hunk 8]
```diff
@@ -16,16 +16,18 @@ export default class NotificationHandler extends Handler {
...
         } catch (e) {
-            utils.handleError(e, res);
+            utils.handleError(e, res, 400);
         }
     }
 
     private async create(req: express.Request, res: express.Response): Promise<void> {
         try {
-            const subscriber = await Subscriber.findOne({id: req.body.subscriberId, userId: req.query.token});
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/notificationHandler.ts]: same as above

File: src/server/handler/notificationHandler.ts [hunk 9]
```diff
@@ -40,7 +42,8 @@ export default class NotificationHandler extends Handler {
 
             notification.id = result.message;
             notification.subscriber = subscriber;
-            notification.userId = req.query.token;
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/notificationHandler.ts]: same as above

File: src/server/handler/spHandler.ts [hunk 10]
```diff
@@ -16,16 +16,18 @@ export default class SpHandler extends Handler {
 
     private async getAll(req: express.Request, res: express.Response): Promise<void> {
         try {
-            const apps = await App.find({userId: req.query.token});
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/spHandler.ts]: why is this here? consider using https://github.com/hopperteam/hopper-server/blob/master/src/handler/authMiddleware.ts#L10-L16

File: src/server/handler/spHandler.ts [hunk 11]
```diff
@@ -16,16 +16,18 @@ export default class SpHandler extends Handler {
...
         } catch (e) {
-            utils.handleError(e, res);
+            utils.handleError(e, res, 400);
         }
     }
 
     private async create(req: express.Request, res: express. Response): Promise<void> {
         try {
-            const response = await this.createApp(req.body, req.query.token.toString(),
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/spHandler.ts]: see above

File: src/server/handler/spHandler.ts [hunk 12]
```diff
@@ -42,7 +44,8 @@ export default class SpHandler extends Handler {
 
     private async update(req: express.Request, res: express. Response): Promise<void> {
         try {
-            const app = await App.findOne({id: req.body.id, userId: req.query.token});
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/spHandler.ts]: see above

File: src/server/handler/subscriberHandler.ts [hunk 13]
```diff
@@ -18,19 +18,22 @@ export default class SubscriberHandler extends Handler {
 
     private async getAll(req: express.Request, res: express.Response): Promise<void> {
         try {
-            const subscriber = await Subscriber.find({userId: req.query.token}).populate("app");
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/subscriberHandler.ts]: why is this here? consider using https://github.com/hopperteam/hopper-server/blob/master/src/handler/authMiddleware.ts#L10-L16

File: src/server/handler/subscriberHandler.ts [hunk 14]
```diff
@@ -18,19 +18,22 @@ export default class SubscriberHandler extends Handler {
...
         } catch (e) {
-            utils.handleError(e, res);
+            utils.handleError(e, res, 400);
         }
     }
 
     private async create(req: express.Request, res: express. Response): Promise<void> {
         try {
-            const app = await App.findOne({id: req.body.appId, userId: req.query.token});
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/subscriberHandler.ts]: see above

File: src/server/handler/subscriberHandler.ts [hunk 15]
```diff
@@ -18,19 +18,22 @@ export default class SubscriberHandler extends Handler {
...
+            const app = await App.findOne({id: req.body.appId, userId: req.session.user.id});
             if (!app)
                 throw new Error("Could not find app");
-            const subscriptionRequest = await subscriberAPI.createSubscriber(req.body, req.query.token.toString(),
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/subscriberHandler.ts]: see above

File: src/server/handler/subscriberHandler.ts [hunk 16]
```diff
@@ -50,7 +53,8 @@ export default class SubscriberHandler extends Handler {
     private async approve(req: express.Request, res: express. Response): Promise<void> {
         try {
             if(!req.query.status.toString().localeCompare("success")){
-                const subscriber = await Subscriber.findOne({_id: req.query.internalId, userId: req.query.token});
+                // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/subscriberHandler.ts]: see above

File: src/server/handler/userHandler.ts [hunk 17]
```diff
@@ -2,41 +2,31 @@ import * as express from "express";
...
-            res.json({
-                "status": "success",
-                "token": user._id
-            });
-        } catch (e) {
-            utils.handleError(e, res);
-        }
+    private async getUser(req: express.Request, res: express.Response): Promise<void> {
+        res.json({
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/userHandler.ts]: why is this here? consider using https://github.com/hopperteam/hopper-server/blob/master/src/handler/authMiddleware.ts#L10-L16

File: src/server/handler/userHandler.ts [hunk 18]
```diff
@@ -2,41 +2,31 @@ import * as express from "express";
...
+            "firstName": req.session.user.firstName,
```
Comment (lucaschimweg) [inline src/server/handler/userHandler.ts:15]: same as above

File: src/server/handler/userHandler.ts [hunk 19]
```diff
@@ -2,41 +2,31 @@ import * as express from "express";
...
+            // @ts-ignore
+            "lastName": req.session.user.lastName,
+            // @ts-ignore
```
Comment (lucaschimweg) [inline src/server/handler/userHandler.ts]: same as above

File: src/server/handler/userHandler.ts [hunk 20]
```diff
@@ -2,41 +2,31 @@ import * as express from "express";
...
-                "status": "success",
-                "token": user._id
-            });
-        } catch (e) {
-            utils.handleError(e, res);
-        }
+    private async logout(req: express.Request, res: express.Response): Promise<void> {
+        res.json({
+            // @ts-ignore
+            "redirect": Config.instance.authRedirectUrl + "?target=" + Config.instance.baseUrl,
```
Comment (lucaschimweg) [inline src/server/handler/userHandler.ts]: this logout won't work. you'll have to delete the cookie (or redirect to hopper-account's logout service which is not implemented yet).

File: src/server/utils.ts [hunk 21]
```diff
@@ -4,10 +4,10 @@ import * as jwt from "jsonwebtoken";
 
 const SALT = "TL]{~eeo=u8J>j>@th8Psh4FQZ:^Wz)UMi;/vXst";
 
-export function handleError(err: Error, res: express.Response, statusCode: number = 400) {
+export function handleError(err: Error, res: express.Response, statusCode: number = 200) {
     console.log(err.message);
```
Comment (lucaschimweg) [inline src/server/utils.ts]: do we really want to print each error? without any metainformation? (uses, accessed resource, ...)

**Review by lucaschimweg — CHANGES_REQUESTED** [review]

File: src/server/handler/userHandler.ts (hunk 20 above)
Comment (konradhartwig) [reply src/server/handler/userHandler.ts]: I just tried to copy the behavior from dev.hoppercloud.net. If you press logout, you got redirect to the account service with a target to dev.hoppercloud.net. But if I go back I'm still logged in so the cookie wasn't deleted either. Is this a wanted behavior?

**Comment by lucaschimweg** [comment]
looks good
